CMake: tests: Support skipping unsupported test with reason

LDong-Arm opened the pull request:
<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html).

NOTE: Do not remove any of the template headings (even for optional sections) as this
template is automatically parsed. 
-->

### Summary of changes <!-- Required -->

Fixes #14856

Add a new argument `TEST_SKIPPED` to `mbed_greentea_add_test()` to indicate a test is skipped and give a reason (e.g. the Mbed target and/or configuration does not provide what the test requires).
    
The skip reason of a test is printed when running tests with `ctest`, and the test is marked as "Skipped" in the test report.

As a showcase, use this mechanism to skip the PSA Attestation test if baremetal is used or PSA and Experimental API are unavailable. Also update the ticker test (whose CTest support was added in #14892) to skip if microsecond ticker is unavailable.

<!-- 
    Please provide the following information: 

    Description of the the change (what is this fixing / adding / removing?).

    Why the change is needed (if this is fixing a reported issue please summarize what
    the issue is and add the reference. E.g. Fixes #17119).

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
    
-->

#### Impact of changes <!-- Optional -->
<!-- 
    If there are any implications for users taking this change then they must be 
    provided here. For Major PR types this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

#### Migration actions required <!-- Optional -->
<!-- 
    This should only be applicable in Major PR types for which this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

### Documentation <!-- Required -->

In-line documentation for `mbed_greentea_add_test()` updated in tools/cmake/mbed_greentea.cmake.

<!-- 
    Please provide details of any document updates required, including links to any
    related PRs against the docs repository.
    If no document updates are required please specify 'None', this at least tells us
    that this has been considered.
-->

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [x] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [] No Tests required for this change (E.g docs only update)
    [] Covered by existing mbed-os tests (Greentea or Unittest)
    [x] Tests / results supplied as part of this PR
    
Got expected behaviors when trying to build the PSA Attestation test in platform/FEATURE_EXPERIMENTAL_API/FEATURE_PSA/TARGET_MBED_PSA_SRV/TESTS/attestation/test,
* with/without `-DMBED_TEST_BAREMETAL=ON`
* with/without `--app-config <path-to>/mbed-os/TESTS/configs/experimental.json`
    
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->

<!--
    Request additional reviewers with @username or @team
-->

@ARMmbed/mbed-os-core 

----------------------------------------------------------------------------------------------------------------

LDong-Arm committed:
cmake: Improve indentation of mbed_greentea_add_test doc

LDong-Arm committed:
CMake: greentea: Port PSA Attestation test to CTest

The PSA Attestation test suite requires full RTOS. There is no need
to explicitly check PSA support, because when we build all greentea
tests from the top of Mbed OS, PSA tests only get included if PSA is
enabled.

LDong-Arm committed:
CMake: tests: Support skipping a test with a reason

Add a new argument `TEST_SKIPPED` to `mbed_greentea_add_test()` to
indicate a test is skipped and give a reason (e.g. the Mbed target
and/or configuration does not provide what the test requires).

The skip reason of a test is printed when running tests with `ctest`,
and the test is marked as "Skipped" in the test report.

LDong-Arm commented on tools/cmake/mbed_greentea.cmake:
```diff
@@ -37,42 +52,53 @@ macro(mbed_greentea_add_test)
+            ALL
+            COMMAND ${CMAKE_COMMAND} -E echo ${msg}
+        )
+    else()
```
For reviewers: Everything below is just indentation...

LDong-Arm commented on platform/FEATURE_EXPERIMENTAL_API/FEATURE_PSA/TARGET_MBED_PSA_SRV/TESTS/attestation/test/CMakeLists.txt:
```diff
@@ -10,9 +10,17 @@ include(${MBED_PATH}/tools/cmake/mbed_greentea.cmake)
 
+if(MBED_TEST_BAREMETAL)
+    set(TEST_SKIPPED "RTOS required")
+elseif(NOT ("PSA" IN_LIST MBED_TARGET_LABELS AND "EXPERIMENTAL_API" IN_LIST MBED_TARGET_LABELS))
```
For now it's here, but when we port this test to CTest like in #14892, this check may not be needed anymore if the directory is not added

LDong-Arm commented on tools/cmake/mbed_greentea.cmake:
```diff
@@ -37,42 +52,53 @@ macro(mbed_greentea_add_test)
+        add_custom_target(${MBED_GREENTEA_TEST_NAME}
+            ALL
+            COMMAND ${CMAKE_COMMAND} -E echo ${msg}
+        )
```
**For discussion**: Do we want to print a skip message?

If we want it, it also makes sense to have its CTest test defined but skipped for consistency. To do that (assuming `add_test()` is called first),
```cmake
set_tests_properties(${MBED_GREENTEA_TEST_NAME}
    PROPERTIES DISABLED
)
```
See tests properties: https://cmake.org/cmake/help/latest/manual/cmake-properties.7.html#properties-on-tests

If we don't want a message, no change is needed in mbed_greentea.cmake - just don't add the test in its own CMakeLists.txt.

LDong-Arm replied:
The idea of printing a "skip" message is to let users know skipped tests have been considered. Keeping discussion open, to see whether we prefer providing more info or simplicity.

rwalton-arm replied:
> **For discussion**: Do we want to print a skip message?
> 
> If we want it, it also makes sense to have its CTest test defined but skipped for consistency. To do that (assuming `add_test()` is called first),
> 
> ```cmake
> set_tests_properties(${MBED_GREENTEA_TEST_NAME}
>     PROPERTIES DISABLED
> )
> ```
> 
> See tests properties: https://cmake.org/cmake/help/latest/manual/cmake-properties.7.html#properties-on-tests
> 
> If we don't want a message, no change is needed in mbed_greentea.cmake - just don't add the test in its own CMakeLists.txt.

I think we should define the skipped test, and set the test properties to report the test as skipped in the CTest test report. This would be useful if we ever wanted to use the CTest report to populate a dashboard view of test results. I think it's more useful to show the test as skipped in the CTest test report than it is to print a "skipped" message at config time.

LDong-Arm replied:
I'll set the test properties when rebasing this PR once #14892 is merged.

I think it doesn't hurt to both print "skipped" at config time and having it in the CTest test report, though the config time message is a bit redundant and can be removed. And there's also a build time message (custom target) in this PR to let the user know why a test is not built.

As for the precise test property to use, I found `DISABLED True` not entirely suitable because it prints

    1/1 Test #1: mbed-drivers-ticker ..............***Not Run (Disabled)  0.00 sec

and the user might ask why it's "Not Run (Disabled)". So I'm thinking of

```cmake
add_test(
    NAME ${MBED_GREENTEA_TEST_NAME}
    COMMAND ${CMAKE_COMMAND} -E echo "Skipping ${MBED_GREENTEA_TEST_NAME}"
}
set_tests_properties(${MBED_GREENTEA_TEST_NAME}
    PROPERTIES
        SKIP_REGULAR_EXPRESSION "."
)
```

which gives

    1/1 Test #1: mbed-drivers-ticker ..............***Skipped   0.01 sec

rwalton-arm commented on tools/cmake/mbed_greentea.cmake:
```diff
@@ -43,42 +52,53 @@ macro(mbed_greentea_add_test)
-    add_subdirectory(${MBED_ROOT} build)
+    if(NOT "${MBED_GREENTEA_TEST_SKIPPED}" STREQUAL "")
+        set(msg "Skipping ${MBED_GREENTEA_TEST_NAME}: ${MBED_GREENTEA_TEST_SKIPPED}")
+        message("${msg}")
```
Is `NOTICE` the correct log level for this message? We could set it to `DEBUG` potentially, as it's probably not that important for a user to see in most cases.

LDong-Arm replied:
I'll change it to debug. The custom target below echoes this message at build stage, which is enough for a user.

LDong-Arm replied:
Done

rwalton-arm reviewed (commented):
Looks pretty good to me. Just one comment about the log level of the skipped message.

Patater commented on tools/cmake/mbed_greentea.cmake:
```diff
@@ -14,8 +14,12 @@ set(MBED_ROOT ${CMAKE_CURRENT_LIST_DIR}/../.. CACHE INTERNAL "")
 # TEST_INCLUDE_DIRS - Test suite include directories for the test
```
In commit message for "CMake: tests: Support skipping unsupported test with reason ":

```diff
- to indicate a test is skip
+ to indicate a test is skipped
```

LDong-Arm replied:
Good catch. Fixed now.

LDong-Arm committed:
CMake: turn mbed_greentea_add_test() into a function

Unlike a macro, a function is more self-contained and can return
early if necessary.

LDong-Arm committed:
CMake: greentea: Add skip reason to ticker test

hazzlim commented on platform/FEATURE_EXPERIMENTAL_API/FEATURE_PSA/TARGET_MBED_PSA_SRV/TESTS/attestation/test/CMakeLists.txt:
```diff
@@ -1,18 +1,19 @@
+if(MBED_GREENTEA_TEST_BAREMETAL)
+    set(TEST_SKIPPED "RTOS required")
+elseif(NOT ("PSA" IN_LIST MBED_TARGET_LABELS AND "EXPERIMENTAL_API" IN_LIST MBED_TARGET_LABELS))
+    set(TEST_SKIPPED "Experiemental API and PSA required")
```
```suggestion
    set(TEST_SKIPPED "Experimental API and PSA required")
```
Tiny typo on a line that sounds like it may not be needed, but just in case!

LDong-Arm replied:
Good point, this line may not be needed. If a users builds test from the top of mbed-os, PSA tests are not even included unless PSA is enabled. This check is only needed when building from the test's own directory which is not recommended, so I'll remove this check.

LDong-Arm replied:
Done

Patater reviewed (approved):
LGTM
